### ISSUE TITLE
Calculate shear correlation function from cosmo model

### ISSUE THREAD

**Issue by dkirkby** [DarkEnergyScienceCollaboration/BremertonRoundTrip#6]
The goal is to calculate the theory prediction for the thing that we estimate from the data.  We already have a tabulated linear theory P(k). Is there already a relatively standalone piece of code to do this?

**Comment by rmjarvis** [comment]
cf. [`galsim.PowerSpectrum.calculateXi()`](https://github.com/GalSim-developers/GalSim/blob/releases/1.3/galsim/lensing_ps.py#L673)

**Comment by rmandelb** [comment]
This calculates the shear correlation function from a shear power spectrum, not from the 3d matter power spectrum (which I think is what they are starting from).

**Comment by rmjarvis** [comment]
Right.  If you haven't done the projection yet, you'll first need to go from P_3d to P_kappa by integrating over the weight function of the sources you are using.

**Comment by dkirkby** [comment]
It looks like `galsim.PowerSpectrum` takes care of going from the shear power spectrum to the correlation function, but we still need to calculate the shear shear power from the background cosmology, DC(z) and DA(z), and 3D matter power spectrum P(k,z).  For concreteness, I want to reproduce the ell = 100 - 1,000 linear-theory curve in this plot for a source redshift of 1:

<img width="683" alt="shearpower" src="https://cloud.githubusercontent.com/assets/185007/9426421/7e507b64-48f4-11e5-8ee6-116389405502.png">

This is taken from [these slides](http://background.uchicago.edu/~whu/Presentations/ferminu.pdf) by Wayne Hu, but unfortunately the cosmology is not specified.  Does anyone know of a better plot to work with?

A notational question for the experts: is there any difference between P_kappa, P_E and C_EE in the flat-sky limit?

**Comment by slosar** [comment]
Yes, as far as I know, P_E=P_EE=P_kappa.
This is the paper I use for the most condensed list of formulae: http://arxiv.org/abs/astro-ph/0112441
I think cosmology hasn't really changed sufficiently,  so that if you get Hu's plot to within 20% you are most likely correct. (i.e. you will miss a factor of pi, not 6/7).

**Comment by slosar** [comment]
I just interfaced treecorr with fastcat. It calculates all 5 correlation functions, but since shear=0, four of them are zero at the moment. 
Treecorr is 2d, so one really calculate wp=int xi(r) along the line of sight. While you are calculate shear corr fnucs, you should also calculate this one with a Gaussian radial window and the corresponding cross-correlation functions.

**Comment by rmjarvis** [comment]
Just FYI, TreeCorr can also do 3d correlation functions if you specify (ra, dec, r) in the input catalog.  It doesn't make much sense for shear, but for counts and scalars, it does the correlations using the direct line physical separation rather than the projected angular separation.

**Comment by dkirkby** [comment]
I just added a [new notebook](https://github.com/DarkEnergyScienceCollaboration/BremertonRoundTrip/blob/master/notebooks/LensingTheoryCalculations.ipynb) showing how to calculate the shear power spectrum using new functions from [randomfield.lensing](http://randomfield.readthedocs.org/en/latest/randomfield/index.html#module-randomfield.lensing).  The plot to compare with Hu's plot above is:

![shearpower](https://cloud.githubusercontent.com/assets/185007/9559809/acdf52c4-4daf-11e5-9d20-1a316aa3df3f.png)

His red dashed (linear) curve should match my green solid curve (z_src = 1), and the agreement looks reasonable given that we aren't using exactly the same cosmology.

@rmjarvis @rmandelb Could the experts look through the plots in the notebook and check that the variations with Omega_k, etc, make sense?

**Comment by dkirkby** [comment]
I added support for calculating cross correlations (shear-shear and shear-galaxy) and for transforming power spectra to correlation functions to [randomfield.lensing](http://randomfield.readthedocs.org/en/latest/randomfield/index.html#module-randomfield.lensing).  There are some new plots at the bottom of this [notebook](https://github.com/DarkEnergyScienceCollaboration/BremertonRoundTrip/blob/master/notebooks/LensingTheoryCalculations.ipynb) to demonstrate these calculations.

For example, this plot shows contributions to shear-shear (top row) and shear-galaxy (bottom row) cross power for three different wavenumbers: ell = 10 (left), 100 (center), 1000 (right).  Average these surfaces over the joint (z1,z2) PDF for a pair of tomographic redshift bins to get the corresponding cross powers at these 3 ell values.  Does this look reasonable?  I am not sure about the sign of the shear-galaxy power: does it look correct here, i.e., Delta**2 > 0 ?

![cross-power](https://cloud.githubusercontent.com/assets/185007/9569213/8a906f26-4f19-11e5-93c4-71d1fe485511.png)

**Comment by slosar** [comment]
Rather than staring into plots this is easiest to debug by brute-force testing in on our pony fields... Should be able to get something soon (but busy next two days).

**Comment by dkirkby** [comment]
I think we need to validate each piece separately, as best we can.  We will do this eventually anyway, when the round trip correlation functions don't agree with the theory, so we might as well start now :-)

For the shear-shear tomography, I tried to reproduce Fig.2 of [Becker 2015, "Cosmic Shear Measurements with DES Science Verification Data"](http://arxiv.org/abs/1507.05598), copied below:
![des-fig2-orig](https://cloud.githubusercontent.com/assets/185007/9572116/8b2e232a-4f5e-11e5-93d7-413965644bca.png)
This involves mocking up the n(z) distributions for the 3 DES photo-z bins, and using them to calculate appropriately weighted sums of the Delta**2(z1,z2,theta) arrays displayed above.  I am just using linear theory, so only go down to 20', but the agreement looks ok to me:
![des-fig2](https://cloud.githubusercontent.com/assets/185007/9572146/0d0a7b3c-4f5f-11e5-9d19-962cab5d7bae.png)
I'm not sure why the DES plot uses such an un-informative vertical scale, but I tried to match it for comparison. There are more details in the usual [notebook](https://github.com/DarkEnergyScienceCollaboration/BremertonRoundTrip/blob/master/notebooks/LensingTheoryCalculations.ipynb).

We still need to validate the shear-galaxy predictions...

**Comment by slosar** [comment]
David, 
i) Please update this on pip if not too hard (otherwise I need to mess with paths, etc.)
ii) It would be good to be able to also predict point correlation functions, both 3D xi(r) (aka good old BAO xi) and projected (wp) while you are at it (I don't want to be pulling this from separate pieces).
iii) Also, the angular matter power spectra for a given window function, both the full calculation, the Limber approx and the thin slice approximation (I can find formulas). Incidentally, this is what we need to trivially do the calculating for Schlegel, so there is a lot of synergy for this...

(I can help you with pieces, rather not since I'm very busy, but so are you, in any case, let's coordinate)
